### PR TITLE
Store logs from 'logexporter' to allow debugging it.

### PR DESCRIPTION
**Pull request by mborsz**
**What this PR does / why we need it**:
With https://github.com/kubernetes/kubernetes/pull/67190 logexporter stopped working properly in 5000 kubemark test (while it works fine in smaller tests)
As we have no tools to debug this in a big scale, I propose storing logexporter logs in some place so that it's possible to debug all attempts.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cluster/log-dump/log-dump.sh [hunk 1]
```diff
@@ -347,6 +347,28 @@ function dump_nodes_with_logexporter() {
     sleep 15
   done
 
+  # Store logs from logexporter pods to allow debugging log exporting process
+  # itself.
+  proc=${max_scp_processes}
+  "${KUBECTL}" get pods -n "${logexporter_namespace}" -o=jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.nodeName}{"\n"}{end}' | while read pod node; do
```
[the remark this answers is not in the document]
Comment (mborsz) [reply cluster/log-dump/log-dump.sh]: Done

File: cluster/log-dump/log-dump.sh [hunk 2]
```diff
@@ -347,6 +347,28 @@ function dump_nodes_with_logexporter() {
     sleep 15
   done
 
+  # Store logs from logexporter pods to allow debugging log exporting process
+  # itself.
+  proc=${max_scp_processes}
```
[the remark this answers is not in the document]
Comment (mborsz) [reply cluster/log-dump/log-dump.sh]: We are not. I just reused that variable. Renamed :)

[the remark this answers is not in the document]
Comment (shyamjvs) [reply cluster/log-dump/log-dump.sh]: Sorry, I meant that you use a different variable here (as max_scp_processes still makes sense at other places in the script where we're SSH'ing). Here I think you can have much higher parallelism (200?) as these operations should be light-weight iiuc.

File: cluster/log-dump/log-dump.sh [hunk 3]
```diff
@@ -347,6 +347,28 @@ function dump_nodes_with_logexporter() {
     sleep 15
   done
 
+  # Store logs from logexporter pods to allow debugging log exporting process
+  # itself.
+  proc=${max_dump_processes}
+  "${KUBECTL}" get pods -n "${logexporter_namespace}" -o jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.nodeName}{"\n"}{end}' | while read pod node; do
+    echo "Fetching logs from ${pod} running on ${node}"
+    mkdir -p ${report_dir}/${node}
+    "${KUBECTL}" logs -n "${logexporter_namespace}" ${pod} > ${report_dir}/${node}/${pod}.txt &
```
Comment (shyamjvs) [inline cluster/log-dump/log-dump.sh]: nit: s/.txt/.log/ (to be consistent)

File: cluster/log-dump/log-dump.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (mborsz) [reply cluster/log-dump/log-dump.sh]: I don't think it's necessary to introduce higher parallelism. Even with 25 concurrent processes here, we should finish dumping data from 5000 nodes cluster in low O(minutes) which should be perfectly sufficient. I would prefer introducing another constant in this file if not necessary.

File: cluster/log-dump/log-dump.sh (hunk 3 above)
Comment (mborsz) [reply cluster/log-dump/log-dump.sh]: Done

**Comment by mborsz** [comment]
https://stackoverflow.com/a/44448420 looks promising, but returns all logs concatenated without any separator nor information where the pod was running.

In most of the cases it should be possible to deduct where the node was running still it requires some tricks -- I would prefer to keep outputs from different nodes separately.

File: cluster/log-dump/log-dump.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (shyamjvs) [reply cluster/log-dump/log-dump.sh]: I'm ok with it, if that's indeed low O(mins) :)

**Comment by mborsz** [comment]
> This should also be be easy to fix. You need to print some log at the start inside logexporter (like Started logexporter pod %v inside node %v)

Yeah, we can change that (in fact even without additional log, it is possible to find out what node is running that component (it prints somewhere the gcs path which contains node name), but I think it shouldn't be a puzzle.

> 
> The reasons I suggested the alternate approach is:
> 
> There is a constant time associated with creating a dir and creating a file in it on GCS. And right now you'd need to do that for 5k files this way
> There is a constant time associated with each kubectl call (for TCP connection, roundtrips with apiserver, etc). And right now we'd make 5k of those (modulo some parallelism)

kubectl logs -l selector is doing O(pods) API calls to fetch logs anyway.

> I'm fine if there's no practical difference in time taken. But we need to be careful to not introduce extra delays (even O(minute)) in our lower-scale presubmits at a time when we're trying to make those faster.

In the presubmit running 100 nodes it added 6 seconds of delay.

I suggest going with the current approach as it's more reliable and easier to use than reading a single log file.
I don't expect this to have significant impact on running time for 5k nodes tests (6 seconds / 100 pods * 5000 pods =~ 5 minutes)

**Comment by mborsz** [comment]
PTAL

**Pull request merged**